### PR TITLE
Facilitate scriptable workspace updates

### PR DESCRIPTION
**Pull request by fnkabit**
API changes required in order to implement #2007 .
Script PR: https://github.com/qownnotes/scripts/pull/120

### CODE REVIEW

File: src/mainwindow.cpp [hunk 1]
```diff
@@ -8596,6 +8605,34 @@ void MainWindow::linkTagNameToCurrentNote(const QString &tagName,
...
+            break;
+        default:
+            windowStateString = "nostate";
+            break;
+        }
+
+        ScriptingService::instance()->callWindowStateChangeHook(windowStateString);
+    }
+
+    QWidget::changeEvent(event);
```
Comment (pbek) [inline src/mainwindow.cpp:8633]: Can you please use QMainWindow::changeEvent, just to make sure.

File: src/mainwindow.h [hunk 2]
```diff
@@ -161,6 +161,8 @@ class MainWindow : public QMainWindow {
     void linkTagNameToCurrentNote(const QString &tagName,
                                   bool linkToSelectedNotes = false);
 
+    void changeEvent(QEvent *event) override;
```
Comment (pbek) [inline src/mainwindow.h:164]: shouldn't that be protected? https://doc.qt.io/qt-5/qwidget.html#changeEvent

File: src/services/scriptingservice.h [hunk 3]
```diff
@@ -46,6 +46,7 @@ class ScriptingService : public QObject {
                                          const QString &action,
                                          const Tag &tag = Tag(),
                                          const QString &newTagName = QString());
+    void callWindowStateChangeHook(const QString &windowStateStr);
```
Comment (pbek) [inline src/services/scriptingservice.h:49]: We'll also need documentation for that hook on the webpage.

File: src/mainwindow.h [hunk 4]
```diff
@@ -188,8 +190,14 @@ class MainWindow : public QMainWindow {
 
     void resetBrokenTagNotesLinkFlag();
 
+    Q_INVOKABLE QString getWorkspaceUuuid(const QString &workspaceName);
```
Comment (pbek) [inline src/mainwindow.h:193]: We'll need documentation for all those on https://www.qownnotes.org/scripting/classes.html#mainwindow

File: src/services/scriptingservice.cpp [hunk 5]
```diff
@@ -496,6 +496,26 @@ QVariant ScriptingService::callNoteTaggingByObjectHook(
     return QVariant();
 }
 
+void ScriptingService::callWindowStateChangeHook(const QString &windowStateStr)
```
Comment (pbek) [inline src/services/scriptingservice.cpp:499]: @fnkabit, @Waqar144 Do you think there ever will be a "StateChangeHook" for another windows than the MainWindow? Would that make sense?

Comment (Waqar144) [reply src/services/scriptingservice.cpp:499]: I think no. I like the idea that we are single window only. A single window is giving you all the abilities. But it also depends on what the user has in mind.

Comment (pbek) [reply src/services/scriptingservice.cpp:499]: Ok, thanks.

File: src/mainwindow.h (hunk 4 above)
Comment (pbek) [reply src/mainwindow.h:191]: see: https://github.com/pbek/QOwnNotes/tree/develop/webpage/src/scripting

Comment (fnkabit) [reply src/mainwindow.h:191]: Updated docs.

Comment (pbek) [reply src/mainwindow.h:191]: Where?

File: src/services/scriptingservice.h (hunk 3 above)
Comment (pbek) [reply src/services/scriptingservice.h:49]: Did you push your documentation commit? I can't see any. 😁

File: src/mainwindow.h (hunk 4 above)
Comment (pbek) [reply src/mainwindow.h:191]: Did you push your documentation commit? I can't see any. 😁

Comment (fnkabit) [reply src/mainwindow.h:191]: Something must have happened with that change. Will redo it :D

Comment (fnkabit) [reply src/mainwindow.h:191]: Forgot to create the PR :)

Comment (pbek) [reply src/mainwindow.h:191]: `getWorkspaceUuuid` has one `u` to many, hasn't it?

Comment (pbek) [reply src/mainwindow.h:191]: I'll refactor that.

**Pull request merged**